### PR TITLE
Implement GC suspension for Linux [WIP]

### PR DESCRIPTION
**Pull request by adityamandaleeka**
This change enables us to redirect or hijack threads that are in cooperative mode when a GC needs to occur.

High-level summary:

- Add a mechanism for injecting activation functions into threads using signals. The function is passed the context that was interrupted.

- Use that mechanism in `SuspendRuntime` in order to inject a function into each thread which has preemptive GC disabled. Using the interrupted context, the function will determine the type of code that was running and react accordingly.

- If we interrupted native code, or if preemptive GC is not disabled for the thread (for instance, if the thread switched back to preemptive mode before the signal handler was invoked), we will exit the signal handler and let the thread continue. 

- If we interrupted managed code, we will use the `ICodeManager` to determine whether the offset we are at is GC safe.

- If we are at a safe point, we push a `RedirectedThreadFrame` on the stack using the captured context, pulse GC mode (enable and disable preemptive GC) for the thread, pop the frame, update the context to reflect modifications by the GC, and exit the signal handler. If we are *not* at a safe point, we will determine whether the method we are in returns a scalar, an object ref, or an interior pointer, and hijack the return address to jump to an appropriate routine which will do the following:

  - Save all the callee-saved registers and the return value on the stack.
  - Call a worker function which will push a hijack frame (which will let us crawl through it back to where the return should have gone), pulse GC mode, and pop the frame.
  - Restore all the captured registers and return to where the function was supposed to return to before the hijack.

### CODE REVIEW

**Comment by kangaroo** [comment]
There doesn't seem to be any guarantee that any function injected with PAL_InjectActivation is signal safe, but it appears to be directly called from the signal handler.  We should probably hijack out of the signal handler immediately and do the work on the target thread outside of a signal context.

cc @sergiy-k

**Comment by janvorli** [comment]
@kangaroo I believe hijacking out of the signal handler would not help. As far as I undestand, the issue with signal safe functions is that you cannot call them because the interrupted thread can easily be executing the same or another one that would collide with the one called from the signal handler. For example due to some internal state.
In our case it cannot happen since we return from the handler right away if we find that the interrupted thread was not executing managed code. And if it was executing managed code, then it couldn't be in any of the problematic native functions.
My point is that hijacking out of the signal handler would not solve anything w.r.t. to the non-signal handler safe functions, since even with the hijack, we would be in the same situation if the interrupted thread was executing a non-signal safe function.

File: src/vm/excep.cpp [hunk 1]
```diff
@@ -7784,11 +7783,19 @@ LONG WINAPI CLRVectoredExceptionHandler(PEXCEPTION_POINTERS pExceptionInfo)
     //
     return retVal;
 #else // !FEATURE_PAL
+
+#if defined(WIN64EXCEPTIONS) && defined(FEATURE_HIJACK)
```
Comment (jkotas) [inline src/vm/excep.cpp]: It may be better to remove the #ifndef FEATURE_PAL instead of duplicating the code.

File: src/vm/threads.cpp [hunk 2]
```diff
@@ -12170,7 +12171,7 @@ HRESULT Thread::YieldTask()
             break;
         }
 
-#ifdef FEATURE_HIJACK
+#if defined(FEATURE_HIJACK) && !defined(PLATFORM_UNIX)
```
Comment (jkotas) [inline src/vm/threads.cpp]: It may be better to put the whole YieldTask method under #ifdef FEATURE_INCLUDE_ALL_INTERFACES.

File: src/vm/threadsuspend.cpp [hunk 3]
```diff
@@ -6853,7 +6880,6 @@ void Thread::HijackThread(VOID *pvHijackAddr, ExecutionState *esb)
 
 #ifdef _DEBUG
     static int  EnterCount = 0;
-    _ASSERTE(EnterCount++ == 0);
```
Comment (jkotas) [inline src/vm/threadsuspend.cpp]: Why did you remove this? It does not make sense now...

**Comment by jkotas** [comment]
@adityamandaleeka take a look at the build breaks...

**Comment by adityamandaleeka** [comment]
@jkotas I looked at the failed build and there are two issues. One was just a missing ifdef, but the other one seems to be that `pthread_sigqueue` isn't available on FreeBSD. I'll investigate what we can do about this.

File: src/vm/threadsuspend.cpp (hunk 3 above)
Comment (adityamandaleeka) [inline src/vm/threadsuspend.cpp]: I removed that assert while debugging, but I'm going to have to change `EnterCount` to not be a static local variable. The reason is that now we can have multiple threads being hijacked at the same time. I'll address this with my next push.

Comment (jkotas) [inline src/vm/threadsuspend.cpp]: I would remove the `EnterCount` variable and all its uses. I do not think that these are a high-value assertions that needs to kept around.

Comment (adityamandaleeka) [inline src/vm/threadsuspend.cpp]: Agreed. I looked at it again and it doesn't seem needed.

File: CMakeLists.txt [hunk 4]
```diff
@@ -522,7 +522,7 @@ if(WIN32)
 endif(WIN32)
 add_definitions(-DFEATURE_EXCEPTIONDISPATCHINFO)
 add_definitions(-DFEATURE_FRAMEWORK_INTERNAL)
-if(WIN32)
+if(WIN32 OR (CLR_CMAKE_PLATFORM_LINUX AND CLR_CMAKE_PLATFORM_UNIX_TARGET_AMD64))
     add_definitions(-DFEATURE_HIJACK)
 endif(WIN32)
```
Comment (adityamandaleeka) [inline CMakeLists.txt]: Fixed endif arguments to match opening.

File: src/vm/threadsuspend.cpp [hunk 5]
```diff
@@ -8231,6 +8247,127 @@ void ThreadSuspend::SuspendEE(SUSPEND_REASON reason)
...
+
+    Thread::WorkingOnThreadContextHolder workingOnThreadContext(pThread);
+    if (!workingOnThreadContext.Acquired())
+        return;
+
+    EECodeInfo codeInfo(ip);
+    if (!codeInfo.IsValid())
+        return;
+
+    DWORD addrOffset = ip - codeInfo.GetStartAddress();
```
Comment (jkotas) [inline src/vm/threadsuspend.cpp]: This should be codeInfo.GetRelOffset() instead.

It is not just a cosmetic fix - there is a real bug. NGen does hot-cold code splitting. The method code can be split into to fragments that are far apart. The reloffset returned by codeInfo for the purpose of GCInfo computation is the logical reloffset as if the hot and cold blocks were next to each other.

File: src/vm/threadsuspend.cpp [hunk 6]
```diff
@@ -8231,6 +8247,127 @@ void ThreadSuspend::SuspendEE(SUSPEND_REASON reason)
...
+        action = pThread->StackWalkFramesEx(
+            &regDisplay,
+            SWCB_GetExecutionState,
+            &executionState,
+            QUICKUNWIND | DISABLE_MISSING_FRAME_DETECTION | ALLOW_ASYNC_STACK_WALK);
+
+        if (action != SWA_ABORT || !executionState.m_IsJIT)
+            return;
+
+        _ASSERTE(executionState.m_ppvRetAddrPtr != NULL);
```
Comment (jkotas) [inline src/vm/threadsuspend.cpp]: I would remove this assert - it is not valid on platforms like ARM where the return address is not guaranteed to be always stored on stack.

**Comment by kangaroo** [comment]
@jasonwilliams200OK pthread_sigqueue also passes the function pointer.  That said, I agree.  Its not a portable implementation.  If we use the tis initial-exec model, we could put the function pointer in a TLS list, which could be accessed by the caller through the CPalThread structure perhaps?

There should be a design which can use pthread_kill which will be portable across all unices.

**Comment by adityamandaleeka** [comment]
@kangaroo @jasonwilliams200OK 

I agree, it's best to have one design that works everywhere. The ifdef I added to `InjectActivationInternal` to check for `pthread_sigqueue` is meant as a temporary band-aid so we can get a clean build and the Linux work is unblocked (since for now, this functionality is only enabled on Linux anyway). It should be removed soon and replaced with a more portable approach.

File: src/vm/threadsuspend.cpp (hunk 5 above)
Comment (adityamandaleeka) [inline src/vm/threadsuspend.cpp]: Thanks, fixed.

**Comment by adityamandaleeka** [comment]
@dotnet-bot retest this please.

**Comment by adityamandaleeka** [comment]
@dotnet-bot retest this please.

File: src/vm/amd64/unixasmhelpers.S [hunk 7]
```diff
@@ -172,6 +172,98 @@ NESTED_ENTRY JIT_RareDisableHelper, _TEXT, NoHandler
 
 NESTED_END JIT_RareDisableHelper, _TEXT
 
+#ifdef FEATURE_HIJACK
+
+//------------------------------------------------
+// OnHijackScalarTripThread
+//
+NESTED_ENTRY OnHijackScalarTripThread, _TEXT, NoHandler
```
Comment (janvorli) [inline src/vm/amd64/unixasmhelpers.S]: I have just realized that with returning structs in registers, we need to add saving restoring rdx as well here. That's the second return value register.

File: src/vm/excep.cpp (hunk 1 above)
Comment (janvorli) [inline src/vm/excep.cpp]: It looks like moving the `#ifndef FEATURE_PAL` below the ` #endif // defined(WIN64EXCEPTIONS)`,  adding `defined(FEATURE_HIJACK)` to the `#if defined(WIN64EXCEPTIONS)` and removing the `#else` would do the trick.

File: src/vm/threadsuspend.cpp [hunk 8]
```diff
@@ -4987,10 +4984,19 @@ HRESULT ThreadSuspend::SuspendRuntime(ThreadSuspend::SUSPEND_REASON reason)
             {
                 FastInterlockOr((ULONG *) &thread->m_State, Thread::TS_GCSuspendPending);
                 countThreads++;
+
+#if defined(FEATURE_HIJACK) && defined(PLATFORM_UNIX)
```
Comment (janvorli) [inline src/vm/threadsuspend.cpp]: Is it correct to have this here? This code is for disabled thread suspension and using polling instead. Why do we inject the suspension here then?

File: src/vm/threadsuspend.cpp [hunk 9]
```diff
@@ -5322,6 +5328,30 @@ HRESULT ThreadSuspend::SuspendRuntime(ThreadSuspend::SUSPEND_REASON reason)
...
+                bool gcSuspensionSignalSuccess = thread->InjectGcSuspension();
+                if (!gcSuspensionSignalSuccess)
+                {
+                    // If we failed to raise the signal for some reason, just log it and move on.
+                    STRESS_LOG1(LF_SYNC, LL_INFO1000, "Thread::SuspendRuntime() -   Failed to raise GC suspension signal for thread %p.\n", thread);
+                }
+            }
+#endif
+
 #ifndef DISABLE_THREADSUSPEND
```
Comment (janvorli) [inline src/vm/threadsuspend.cpp]: It seems it would be nice to merge the loop above and below and just have the actual suspension mechanism ifdefed. And maybe even better, refactor the code to extract the suspension mechanism into a separate function(s).

File: src/vm/threadsuspend.cpp [hunk 10]
```diff
@@ -8231,6 +8247,126 @@ void ThreadSuspend::SuspendEE(SUSPEND_REASON reason)
...
+    }
+    else
+    {
+        // The thread is in non-interruptible code.
+        ExecutionState executionState;
+        StackWalkAction action;
+        REGDISPLAY regDisplay;
+        pThread->InitRegDisplay(&regDisplay, interruptedContext, true /* validContext */);
+
+        if (!pThread->IsSafeToInjectThreadAbort(interruptedContext))
```
Comment (janvorli) [inline src/vm/threadsuspend.cpp]: Why is this needed here? We are not trying to inject thread abort, right?

File: src/vm/threadsuspend.cpp [hunk 11]
```diff
@@ -8231,6 +8247,126 @@ void ThreadSuspend::SuspendEE(SUSPEND_REASON reason)
...
+            &executionState,
+            QUICKUNWIND | DISABLE_MISSING_FRAME_DETECTION | ALLOW_ASYNC_STACK_WALK);
+
+        if (action != SWA_ABORT || !executionState.m_IsJIT)
+            return;
+
+        if (executionState.m_ppvRetAddrPtr == NULL)
+            return;
+
+        ENABLE_FORBID_GC_LOADER_USE_IN_THIS_SCOPE();
```
Comment (janvorli) [inline src/vm/threadsuspend.cpp]: Can you add a comment why this is needed? I assume it is in order to make sure that the GetMethodDesc below doesn't throw or trigger a GC, based on the comment on the macro, but I am not sure.

File: src/vm/threadsuspend.cpp [hunk 12]
```diff
@@ -8231,6 +8247,126 @@ void ThreadSuspend::SuspendEE(SUSPEND_REASON reason)
...
+        // Mark that we are performing a stackwalker like operation on the current thread.
+        // This is necessary to allow the signature parsing functions to work without triggering any loads.
+        ClrFlsValueSwitch _threadStackWalking(TlsIdx_StackWalkerWalkingThread, pThread);
```
Comment (janvorli) [inline src/vm/threadsuspend.cpp]: Why does the local name start with the underscore?

**Comment by janvorli** [comment]
I also wonder - was the shuffle of large pieces of code in the threadsuspend.cpp necessary? The diff got crazy about it and I had to get the file content to a local file, diff it using a merge tool against the current version and reorder it back to be able to see that there were no changes in the methods.

**Comment by adityamandaleeka** [comment]
@janvorli That shuffle was done to merge non-contiguous chunks of code which had the same ifdefs. With the latest changes made to the ifdefs, I'm not sure how much of it is still necessary, but yes, the moved functions are unchanged.

File: src/vm/threadsuspend.cpp (hunk 8 above)
Comment (adityamandaleeka) [inline src/vm/threadsuspend.cpp]: Looks like the comment above (about using GCPOLL) should be updated since now we have a scenario where thread suspension is disabled but we do something other than just rely on polling.

File: src/vm/threadsuspend.cpp (hunk 12 above)
Comment (adityamandaleeka) [inline src/vm/threadsuspend.cpp]: Will fix. Copied it from another place in the file :)

File: src/vm/excep.cpp (hunk 1 above)
Comment (adityamandaleeka) [inline src/vm/excep.cpp]: Done. This means that block above with the logging will also run when FEATURE_PAL is enabled now.

File: src/vm/threadsuspend.cpp (hunk 11 above)
Comment (adityamandaleeka) [inline src/vm/threadsuspend.cpp]: Sure, will add a comment. I believe this macro doesn't actually impact anything in release mode. Its purpose is to turn off the GC_TRIGGERS/THROWS/INJECT_FAULT contract in LoadTypeHandle in cases where the caller won't try to load a type. I included it here because it's done in the Windows code.

File: src/vm/excep.cpp (hunk 1 above)
Comment (janvorli) [inline src/vm/excep.cpp]: This doesn't look like what I meant. Is there a way to see the previous version somehow so that I can check what  was what I really was suggesting?

Comment (janvorli) [inline src/vm/excep.cpp]: I am sorry, I have confused myself into thinking the change is already visible in this diff.

File: src/vm/amd64/unixasmhelpers.S (hunk 7 above)
Comment (adityamandaleeka) [inline src/vm/amd64/unixasmhelpers.S]: Done.

Comment (adityamandaleeka) [inline src/vm/amd64/unixasmhelpers.S]: Done in my latest commit, but I now need to update HijackArgs to include rdx. Will update it soon.

Comment (adityamandaleeka) [inline src/vm/amd64/unixasmhelpers.S]: Actually, I might not need to do this, if I just make RDI point to where the HijackArgs are.

Comment (adityamandaleeka) [inline src/vm/amd64/unixasmhelpers.S]: Spoke to @janvorli about this. We'll wait on doing this until after the work for passing structs in registers is ready to be merged. Once we do, we'll also need to ensure that object refs in the registers are handled properly.

Comment (kangaroo) [inline src/vm/amd64/unixasmhelpers.S]: Dear everyone @microsoft (cc @richlander) for public PR's can we PLEASE stop having "I spoke with @othermicrosoftie about this" as a response?

*sigh*

Comment (adityamandaleeka) [inline src/vm/amd64/unixasmhelpers.S]: @kangaroo I've updated the comment to remove the part about speaking offline. Sorry about that. However, the rest of my comment stands on its own as a response. The work for passing structs in registers hasn't been merged yet (it is in a public fork), which is why we decided not to address it now for this PR.

**Comment by jkotas** [comment]
Could you please add a config switch (src\inc\clrconfigvalues.h) that can be used to turn the thread hijacking on/off. The config switch should be on by default, when turned off - skip calls to PAL_InjectActivation.

I expect this config switch will be useful to quickly rule out theories on whether a crash is caused by thread hijacking.

**Comment by adityamandaleeka** [comment]
@jkotas Good idea. I've added a config switch called `INTERNAL_GCDisableUnixActivationInjection`, and it will be including in my next push.

**Comment by jkotas** [comment]
@adityamandaleeka I think it is a bad practice to name config switches "DisableABC". I know that there are quite a few - but it does not make them good. It leads to double negative that one needs to think twice about. I think it is better to make the config switch turned on by default instead (there are quite a few of these as well).

it would be nice to make the name mention thread suspend for discoverability.

`INTERNAL_ThreadSuspendInjection` is the best name I was able to come up with.

**Comment by adityamandaleeka** [comment]
@jkotas Sounds good. I noticed that almost all of the GC-related switches are prefaced with 'GC'. How about `INTERNAL_GCThreadSuspendInjection` (or should this be in a different section of clrconfigvalues.h)?

**Comment by jkotas** [comment]
The thread suspension is implemented as a shared service in CoreCLR - used by GC, debugger, profiler, ... . The existing suspend related config switches do not have GC prefix, e.g. INTERNAL_SuspendDeadlockTimeout

**Comment by adityamandaleeka** [comment]
Done. I created a section called "Thread Suspend" for the new switch and moved some stuff from "Unknown" to there as well.

**Comment by adityamandaleeka** [comment]
Thanks @jkotas, updated.

**Comment by jkotas** [comment]
LGTM

**Comment by jkotas** [comment]
Are you still waiting for something before it is merged in? I think it is good to go.

**Comment by adityamandaleeka** [comment]
@jkotas I will merge it soon. I've been testing this work some more and discovered an infrequent crash under GCStress that, so far, seems to indicate that our epilog detection is a bit off sometimes on Linux. I'd like to investigate more and fix it before merging (or at least open an issue if it's out of scope).

File: src/vm/threadsuspend.cpp (hunk 10 above)
Comment (adityamandaleeka) [inline src/vm/threadsuspend.cpp]: I've replaced this with a call to IsIPInEpilog (which was the real intention behind calling this here anyway).
